src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+    ///
+    /// epoch::pin(|scope| {
+    ///     let mut curr = a.load(SeqCst, scope);
+    ///     let res = a.compare_and_swap_owned(curr, Owned::new(5678), SeqCst, scope);
+    /// });
+    /// ```
+    pub fn compare_and_swap_owned<'scope>(
+        &self,
+        current: Ptr<T>,
+        new: Owned<T>,
+        ord: Ordering,
+        _: &'scope Scope,
+    ) -> Result<Ptr<'scope, T>, (Ptr<'scope, T>, Owned<T>)> {
+        let fail_ord = strongest_failure_ordering(ord);
+        match self.data.compare_exchange(current.data, new.data, ord, fail_ord) {
+            Ok(_) => {
+                let data = new.data;
+                mem::forget(new);
+                Ok(Ptr::from_data(data))
+            }
+            Err(previous) => Err((Ptr::from_data(previous), new)),
+        }
+    }
+
+    /// Stores `new` into the atomic pointer if the current value is the same as `current`.
+    ///
+    /// Unlike [`compare_and_swap_owned`], this method is allowed to spuriously fail even when
+    /// comparison succeeds, which can result in more efficient code on some platforms.
+    /// The return value is a result indicating whether the new pointer was written. On success the
+    /// pointer that was written is returned. On failure `new` and the actual current value are
+    /// returned.
+    ///
+    /// This method takes an [`Ordering`] argument which describes the memory ordering of this
+    /// operation.
+    ///
+    /// [`compare_and_swap_owned`]: struct.Atomic.html#method.compare_and_swap_owned
+    /// [`Ordering`]: https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_epoch::{self as epoch, Atomic, Owned};
+    /// use std::sync::atomic::Ordering::SeqCst;
+    ///
+    /// let a = Atomic::new(1234);
+    ///
+    /// epoch::pin(|scope| {
+    ///     let mut new = Owned::new(5678);
+    ///     let mut ptr = a.load(SeqCst, scope);
+    ///     loop {
+    ///         match a.compare_and_swap_weak_owned(ptr, new, SeqCst, scope) {
+    ///             Ok(p) => {
+    ///                 ptr = p;
+    ///                 break;
+    ///             }
+    ///             Err((p, n)) => {
+    ///                 ptr = p;
+    ///                 new = n;
+    ///             }
+    ///         }
+    ///     }
+    /// });
+    /// ```
+    pub fn compare_and_swap_weak_owned<'scope>(
+        &self,
+        current: Ptr<T>,
+        new: Owned<T>,
+        ord: Ordering,
+        _: &'scope Scope,
+    ) -> Result<Ptr<'scope, T>, (Ptr<'scope, T>, Owned<T>)> {
+        let fail_ord = strongest_failure_ordering(ord);
+        match self.data.compare_exchange_weak(current.data, new.data, ord, fail_ord) {
+            Ok(_) => {
+                let data = new.data;
+                mem::forget(new);
+                Ok(Ptr::from_data(data))
+            }
+            Err(previous) => Err((Ptr::from_data(previous), new)),
+        }
+    }
+}
```

Comment:
As @Vtec234 pointed out, it would be nice to implement `fetch_??` series (as defined in stdlib).